src/net/udp.rs

```diff
@@ -243,12 +283,66 @@ impl UdpSocket {
     }
 }
 
+
 impl fmt::Debug for UdpSocket {
     fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
         self.io.get_ref().fmt(f)
     }
 }
 
+/// A future used to write the entire contents of some data to a stream.
+///
+/// This is created by the [`write_all`] top-level method.
+///
+/// [`write_all`]: fn.write_all.html
+pub struct SendAllTo<'a, T> {
+    state: UdpState<'a, T>,
+}
+
+enum UdpState<'a, T> {
+    Writing {
+        sock: &'a UdpSocket,
+        buf: T,
+        addr: &'a SocketAddr,
+        pos: usize,
+    },
+    Empty,
+}
+
+
+fn zero_write() -> io::Error {
+    io::Error::new(io::ErrorKind::WriteZero, "zero-length write")
+}
+
+impl<'a, T> Future for SendAllTo<'a, T>
+    where T: AsRef<[u8]>,
+{
+    type Item = T;
+    type Error = io::Error;
+
+    fn poll(&mut self) -> Poll<T, io::Error> {
+        match self.state {
+            UdpState::Writing { ref sock, ref buf, ref addr, ref mut pos } => {
+                let buf = buf.as_ref();
+                while *pos < buf.len() {
```

Comment:
You're right.  This will transparently send multiple datagrams.  
We should probably remove the loop and then verify that the the buf doesn't exceed a max datagram size, and probably return an error if it didn't send buf size, as the datagram would be smaller than intended.